docs/rfcs/2023-03-08-region-fault-tolerance.md

```diff
@@ -0,0 +1,194 @@
+
+To conclude, the failure detection pseudo-codes are like this:
+
+```rust
+// in Meta:
+fn failure_detection() {
+    loop {
+        // passive detection
+        let failed_regions = all_regions.iter().filter(|r| r.estimated_failure_possibility() > config.phi).collect();
+
+        // find the datanodes that contains the failed regions
+        let datanodes_and_regions = find_region_resides_datanodes(failed_regions);
+
+        // active detection  
+        for (datanode, regions) in datanodes_and_regions {
+            if !datanode.is_healthy(regions) {
+                do_failover(datanode, regions);
+            }
+        }
+
+        sleep(config.detect_interval);
+    }
+}
+```
+
+Some design considerations:
+
+- Why active detecting while we have passively detection? Because it could be happened that the network is singly connectable sometimes (especially in the complex Cloud environment), then the Datanode's heartbeats cannot reach Meta, while Meta could request Datanode. Active detecting avoid this false positive situation.
+- Why use region heartbeat instead of Datanode heartbeat, i.e. the detection works on region instead of Datanode? Because we might face the possibility that only part of the regions in the Datanode are not available, not ALL regions. Especially the situation that Datanodes are used by multiple tenants. If this is the case, it's better to do failover upon the designated regions instead of the whole regions that reside on the Datanode. All in all, we want a more subtle control over region failover. 
+
+So we detect some regions are not available. How to regain the availability back?
+
+## Region Failover
+
+Region Failover largely relies on remote WAL, aka "[Bunshin](https://github.com/GreptimeTeam/bunshin)". I'm not including any of the details of it in this RFC, let's just assume we already have it.
+
+In general, region failover is fairly simple. Once Meta decides to do failover upon some regions, it first chooses one or more Datanodes to hold the failed region. This can be done easily, as the Meta already has the whole picture of Datanodes: it knows which Datanode has the minimum regions, what Datanode historically had the lowest CPU usage and IO rate, and how the Datanodes are assigned to tenants, among other information that can all help the Meta choose the most suitable Datanodes. Let's call these chosen Datanodes as "candidates".
+
+Then, Meta sets the states of these failed regions as "passive". We should add a field to `Region`:
+
+```protobuf
+message Region {
+  uint64 id = 1;
+  string name = 2;
+  Partition partition = 3;
+  
+  message State {
+    Active,
+    Passive,
+  }
+  State state = 4;
+  
+  map<string, string> attrs = 100;
+}
+```
+
+Here `Region` is used in message `RegionRoute`, which indicates how the write request is split among regions. When a region is set as "passive", Frontend knows the write to it should be rejected at the moment (the region read is not blocked, however).
```

Comment:
How does the Frontend know that a certain Datanode node is unavailable? There may be several ways:

1. Rely on Frontend to obtain new information from metasrv after cache invalidation

2. Metasrv actively pushes messages indicating that a datanode is unavailable

If option 1 is chosen, long cache TTL can lead to high latency, while reducing the cache TTL will increase pressure on metasrv.

Option 2 requires Frontend to actively establish a connection with Metasrv and wait for it to push messages.

Reply:
I prefer option 2. Let's do it in Frontend's heartbeat to Metasrv.